A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+trailing metadata of the server call and send back with the RPC response.
+
+Define a metric recorder that is unique per RPC. Server applications should obtain a reference and 
+use it to record metrics specifically for the current RPC call. Implementing this may leverage a 
+[context](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html) or similar mechanism, 
+and a language should implement one if not already existed.
+The metric recoder looks like this:
+
+```Java
+public final class CallMetricRecorder {
+  // Records a request cost metric measurement for the call.
+  public CallMetricRecorder recordRequestCostMetric(String name, double value);
+
+  // Records a utilization metric measurement for the call.
+  public CallMetricRecorder recordUtilizationMetric(String name, double value);
+
+  //Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordCPUUtilizationMetric(double value);
+
+  // Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordMemoryUtilizationMetric(double value);
+
+  // Returns the call metric recorder attached to the current call.
+  public static CallMetricRecorder getCurrent();
+}
+```
+
+All the methods should be thread safe.
+Recording one metrics multiple times overrides the previously provided values. 
+The methods can be called at any time during an RPC lifecycle,
+depending on the particular type of the cost or utilization measurement. For example, an RPC request size may
+be recorded at the beginning of the call, while a queue-size may be reported at any time during the call life-cycle.
+The data is dumped and translated to `OrcaLoadReport` at the end of each RPC by the interceptor, 
+and sent out as the trailing metadata.
+
+Note that reporting utilization metrics per-query is only useful in high QPS systems. 
+In such systems, the application that measures utilization data should consider sampling the data 
+to avoid performance degradation due to high rates of syscalls.
+
+
+### Out of Band Metrics Reporting
+For the out-of-band reporting, it requires a separate load reporting server running along with the 
+backend server for as long as the server is alive.
+
+At the client side, typically an LB policy, opens a stream on each
+established connection to the backend server to request the load report, and then periodically receiving
+metrics data from the streaming server for each backend.
+The report request should specify metrics reporting interval, the lower bound of which should be capped 
+to a server configurable value, and there is no upper bound unless it is needed in the future.
+
+#### The Server API 
+Provide builtin implementation of `OpenRCAService` streaming service defined by ORCA in each supported language.
+Server applications are free to add such a load reporting server
+in a normal way that they register a service.
+
+The service holds the utilization metrics data in key/value pairs using a hashmap. 
```

Comment:
I don't think we should over-specify the implementation here; the type of map to use is an implementation detail that this design doesn't really have to care about.  I think it's sufficient to say that it's a map.